File: src/logic/querys.js
const { getValue } = require("./conditions");

function getQuery(condition, userData = {}) {
    let upper = 0;
    let lower = 0;
    let value = '';
    if (condition.operator == 'BETWEEN') {
        let values = condition.value.split(',');
        upper = getCorrectValueType(values[1], condition.valueType);
        lower = getCorrectValueType(values[0], condition.valueType);
    } else {
        value = getValue(condition.source, condition.value, condition.valueType, userData);
    }

    if (condition.valueType == "BOOLEAN") {
        value = value == true ? "true" : "false";
    }

    let queryValue = condition.valueType == "STRING" ? `"${value}"` : value;
    switch (condition.operator) {
        case "EQUALS":
        case "==":
            return `{"${condition.field}": ${queryValue}}`;
        case "IN":
        case "in":
            return `{"${condition.field}": {"$in": [${queryValue}]}}`;
        case "NOT_EQUALS":
        case "!=":
            return `{"${condition.field}": {"$ne": ${queryValue}}}`;
        case "GREATER_THAN":
        case ">":
            return `{"${condition.field}": {"$gt": ${queryValue}}}`;
        case "LESS_THAN":
        case "<":
            return `{"${condition.field}": {"$lt": ${queryValue}}}`;
        case "GREATER_THAN_OR_EQUAL":
        case ">=":
            return `{"${condition.field}": {"$gte": ${queryValue}}}`;
        case "LESS_THAN_OR_EQUAL":
        case "<=":
            return `{"${condition.field}": {"$lte": ${queryValue}}}`;
        case "BETWEEN":
            return `{"${condition.field}": {"$gte": ${lower}, "$lte": ${upper}}}`;
        default:
            return `{"${condition.field}": ${queryValue}}`;
    }
}

function getQueryDynamo(condition, userData = {}) {
    let upper = 0;
    let lower = 0;
    let value = '';
    if (condition.operator == 'BETWEEN') {
        let values = condition.value.split(',');
        upper = getCorrectValueType(values[1], condition.valueType);
        lower = getCorrectValueType(values[0], condition.valueType);
    } else {
        value = getValue(condition.source, condition.value, condition.valueType, userData);
    }

    if (condition.valueType == "BOOLEAN") {
        value = value == true ? "true" : "false";
    }

    let queryValue = condition.valueType == "STRING" ? `${value}` : value;


    switch (condition.operator) {
        case "BETWEEN":
            return `${lower} AND ${upper}`;
        default:
            return queryValue;
    }
}

function getOperatorDynamo(operator) {
    switch (operator) {
        case "EQUALS":
        case "==":
            return `=`;
        case "NOT_EQUALS":
        case "!=":
            return `<>`;
        case "GREATER_THAN":
        case ">":
            return `>`;
        case "LESS_THAN":
        case "<":
            return `<`;
        case "GREATER_THAN_OR_EQUAL":
        case ">=":
            return `>=`;
        case "LESS_THAN_OR_EQUAL":
        case "<=":
            return `<=`;
        case "BETWEEN":
            return `BETWEEN`;
        default:
            return `=`;
    }
}

module.exports = { getQuery, getQueryDynamo ,getOperatorDynamo}

File: src/py_logic/querys.py
def get_query_dynamo(condition, user_data={}):
    upper = 0
    lower = 0
    value = ''
    if condition['operator'] == 'BETWEEN':
        values = condition['value'].split(',')
        upper = get_correct_value_type(values[1], condition['valueType'])
        lower = get_correct_value_type(values[0], condition['valueType'])
    else:
        value = get_value(condition['source'], condition['value'], condition['valueType'], user_data)

    if condition['valueType'] == "BOOLEAN":
        value = "true" if value else "false"

    query_value = str(value) if condition['valueType'] == "STRING" else value

    operator = condition['operator']

    if operator == "BETWEEN":
        return f"{lower} AND {upper}"
    else:
        return query_value

def get_operator_dynamo(operator):
    if operator in ["EQUALS", "=="]:
        return "="
    elif operator in ["NOT_EQUALS", "!="]:
        return "<>"
    elif operator in ["GREATER_THAN", ">"]:
        return ">"
    elif operator in ["LESS_THAN", "<"]:
        return "<"
    elif operator in ["GREATER_THAN_OR_EQUAL", ">="]:
        return ">="
    elif operator in ["LESS_THAN_OR_EQUAL", "<="]:
        return "<="
    elif operator == "BETWEEN":
        return "BETWEEN"
    else:
        return "="
